### PR TITLE
faster begintimestep

### PR DESCRIPTION
**Pull request by hnil**
- faster updateProperty
- refactored for making local updating in inhereted classes

### CODE REVIEW

**Comment by bska** [comment]
benchmark please

**Comment by bska** [comment]
jenkins build this please

File: ebos/eclproblem.hh [hunk 1]
```diff
@@ -2183,19 +2206,30 @@ private:
...
         return false;
     }
 
+    bool updateMaxOilSaturation_(unsigned compressedDofIdx, const IntensiveQuantities& iq) 
+    {
+        OPM_TIMEBLOCK_LOCAL(updateMaxOilSaturation);
+        const auto& fs = iq.fluidState();
+        const Scalar So = decay<Scalar>(fs.saturation(oilPhaseIdx));
+        auto& mos = this->maxOilSaturation_;
+        if(mos[compressedDofIdx], So){
```
Comment (akva2) [inline ebos/eclproblem.hh:2225]: This looks wrong.

Comment (hnil) [reply ebos/eclproblem.hh]: fixed.

**Comment by bska** [comment]
jenkins build this please

**Comment by ytelses** [comment]
Benchmark result overview: 




| Test    | Configuration                              | Relative |
| ------- | ------------------------------------------ | -------- |
| opm-git | OPM Benchmark: drogon - Threads: 1         | 1.02     |
| opm-git | OPM Benchmark: drogon - Threads: 8         | 1.01     |
| opm-git | OPM Benchmark: smeaheia - Threads: 1       | 1.01     |
| opm-git | OPM Benchmark: smeaheia - Threads: 8       | 0.976    |
| opm-git | OPM Benchmark: spe10_model_1 - Threads: 1  | 0.983    |
| opm-git | OPM Benchmark: spe10_model_1 - Threads: 8  | 0.999    |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 1 | 0.999    |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 8 | 0.864    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 1 | 0.885    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 8 | 0.829    |
* Speed-up = Total time master / Total time pull request. Above 1.0 is an improvement. *



View result details @ https://www.ytelses.com/opm/?page=result&id=2012

**Comment by ytelses** [comment]
Benchmark result overview: 




| Test    | Configuration                              | Relative |
| ------- | ------------------------------------------ | -------- |
| opm-git | OPM Benchmark: drogon - Threads: 1         | 1.023    |
| opm-git | OPM Benchmark: drogon - Threads: 8         | 1.003    |
| opm-git | OPM Benchmark: smeaheia - Threads: 1       | 0.999    |
| opm-git | OPM Benchmark: smeaheia - Threads: 8       | 0.977    |
| opm-git | OPM Benchmark: spe10_model_1 - Threads: 1  | 0.998    |
| opm-git | OPM Benchmark: spe10_model_1 - Threads: 8  | 1        |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 1 | 0.996    |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 8 | 0.89     |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 1 | 1.014    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 8 | 1        |
* Speed-up = Total time master / Total time pull request. Above 1.0 is an improvement. *



View result details @ https://www.ytelses.com/opm/?page=result&id=2012

**Comment by hnil** [comment]
Fixed issue with use of on the fly intensive quantities in the at start of simulation. May it be similar situation for restart cases which is not taken by tests?, Most relevant restart where explicit variables is used in the simulation (i.e. maxXXX + hyster). All tests passed local now.

**Comment by bska** [comment]
jenkins build this please

File: ebos/eclproblem.hh [hunk 2]
```diff
@@ -2081,24 +2068,50 @@ public:
...
     }
 
-private:
     template<class UpdateFunc>
     void updateProperty_(const std::string& failureMsg,
                          UpdateFunc func)
     {
         OPM_TIMEBLOCK(updateProperty);
-        ElementContext elemCtx(this->simulator());
+        const auto& problem = this->simulator().problem();
```
Comment (bska) [inline ebos/eclproblem.hh:2103]: Tiny nit: This `problem` object is not used anywhere in the `updateProperty_()` function.  Is that intentional?  If so, I suggest we just remove it.

**Review by bska — APPROVED** [review]
I like it, especially the part that moves the bodies of some large lambdas out to separate helper functions.  Apart from one small nit I think I'd like to merge this now.

File: ebos/eclproblem.hh (hunk 2 above)
Comment (bska) [reply ebos/eclproblem.hh]: > fixed

Excellent.  I'll just run final build check and then merge when the check is green.

**Comment by bska** [comment]
jenkins build this please

**Comment by bska** [comment]
> I'll just run final build check and then merge when the check is green.

The build was successful.  I'll merge into master now.

**Pull request merged**